platforms/avr/drivers/ws2812_spi.c

```diff
@@ -0,0 +1,64 @@
+/*
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This is a SPI driver for the SK6812/WS2812 RGB leds,
+ * for the ATmega32U.
+ *
+ * It sets SPI to 8MHz and sends 11100000 for a zero,
+ * or 11111110 for a one. The extra time to prepare the
+ * next bit ensures the timming after each bit.
+ *
+ * It requires a 16MHz CPU clock and the pins PB0 and PB1
+ * unused (they are part of the ISP interface, so it's not
+ * uncommon), and the leds connected to the PB2 pin. Also,
```

Comment:
My fault: it wasn't PB0 but PB3. Fixed. Anyway, those boards can use the bit-banging code.

Reply:
It's all of PB0…PB3 — PB0 is the SS signal, and that pin must be configured as output for the SPI master to work properly (if PB0 is configured as input, and that pin reads as low, the SPI controller switches from the master to the slave mode).

So you have these restrictions on pins:
- PB0 — must be configured as output, should be usable as a plain output pin; reconfiguring this pin as input will break the SPI master functionality;
- PB1 — not sure (the datasheet says that it can be reconfigured as input — maybe that means that it would be usable as a plain input pin; but if that pin is configured as output, the SPI controller will override the pin state to output SCK);
- PB2 — output pin for the WS2812 data;
- PB3 — configured as input if SPI master is enabled; should be usable as a plain input pin.

Reply:
You are right. Fixed.